### PR TITLE
List rule changes in ecosystem

### PR DESCRIPTION
**Pull request by calumy**
Currently, the ecosystem check lists all changes between two versions of ruff as a diff. This PR counts the changes for each rule so the impact of the change can be determined. 

A use case where this could be useful is when adding a set of new rules to determine if one rule will cause a particularly large number of new rule violations to be raised. For example, the addition of the new flake8-todo rules in #3921 raised [a discussion ](https://github.com/charliermarsh/ruff/pull/3921#discussion_r1190257550)about the impact of the rule TDO003 which checks that TODOs have links present.

## Example output

When comparing the current build to v0.0.264 the example output would be.

```bash
scripts/check_ecosystem.py env/bin/ruff target/release/ruff
```

ℹ️ ecosystem check **detected changes**. (+12, -0, 0 error(s))

<details><summary>typeshed (+12, -0)</summary>
<p>

```diff
+ stdlib/typing.pyi:735:1: PYI042 Type alias `_get_type_hints_obj_allowed_types` should be CamelCase
+ stubs/cffi/_cffi_backend.pyi:113:5: PYI042 Type alias `buffer` should be CamelCase
+ stubs/cffi/cffi/api.pyi:13:1: PYI042 Type alias `basestring` should be CamelCase
+ stubs/cffi/cffi/api.pyi:18:5: PYI042 Type alias `buffer` should be CamelCase
+ stubs/pika/pika/spec.pyi:9:1: PYI042 Type alias `_str` should be CamelCase
+ stubs/pywin32/pythoncom.pyi:10:1: PYI042 Type alias `error` should be CamelCase
+ stubs/pywin32/win32/odbc.pyi:9:1: PYI042 Type alias `_odbcError` should be CamelCase
+ stubs/pywin32/win32comext/adsi/adsi.pyi:7:1: PYI042 Type alias `error` should be CamelCase
+ stubs/pywin32/win32comext/propsys/propsys.pyi:6:1: PYI042 Type alias `error` should be CamelCase
+ stubs/pywin32/win32comext/shell/shell.pyi:7:1: PYI042 Type alias `error` should be CamelCase
+ stubs/redis/redis/typing.pyi:17:1: PYI043 Private type alias `_StringLikeT` should not be suffixed with `T` (the `T` suffix implies that an object is a `TypeVar`)
+ stubs/requests/requests/compat.pyi:22:1: PYI042 Type alias `builtin_str` should be CamelCase
```

</p>
</details>
Rules changed: 2

| Rule | Changes | Additions | Removals |
| ---- | ------- | --------- | -------- |
| PYI042 | 11 | 11 | 0 |
| PYI043 | 1 | 1 | 0 |

### CODE REVIEW

File: scripts/check_ecosystem.py [hunk 1]
```diff
@@ -294,9 +295,47 @@ async def main(*, ruff1: Path, ruff2: Path, projects_jsonl: Optional[Path]) -> N
...
+                    elif line[0] == "-":
+                        current_changes = (current_changes[0], current_changes[1] + 1)
+
+                    rule_changes[rule_code] = current_changes
+
             else:
                 continue
 
+        print(f"Rules changed: {len(rule_changes.keys())}")
+        print()
```
Comment (calumy) [inline scripts/check_ecosystem.py:327]: I'm not sure this is adding anything so could be removed

File: scripts/check_ecosystem.py [hunk 2]
```diff
@@ -294,9 +295,47 @@ async def main(*, ruff1: Path, ruff2: Path, projects_jsonl: Optional[Path]) -> N
                 print()
                 print("</p>")
                 print("</details>")
+
+                # Count rule changes
+                for line in diff_str.splitlines():
+                    # Find rule change for current line or construction
+                    # + <rule>/<path>:<line>:<column>: <rule_code> <message>
+                    matches = re.findall(r": [A-Z]{1,3}[0-9]{3,4}", line)
```
Comment (konstin) [inline scripts/check_ecosystem.py:303]: just finding the first match is sufficient here

File: scripts/check_ecosystem.py [hunk 3]
```diff
@@ -294,9 +295,47 @@ async def main(*, ruff1: Path, ruff2: Path, projects_jsonl: Optional[Path]) -> N
...
+                    if len(matches) == 0:
+                        # Handle case where there are no regex matches e.g.
+                        # +                 "?application=AIRFLOW&authenticator=TEST_AUTH&role=TEST_ROLE&warehouse=TEST_WAREHOUSE" # noqa: E501, ERA001
+                        # Which was found in local testing
+                        continue
+
+                    rule_code = matches[0][2:]  # Trim leading ": "
```
Comment (konstin) [inline scripts/check_ecosystem.py:310]: i'd use a [match group](https://docs.python.org/3/howto/regex.html#grouping) here

File: scripts/check_ecosystem.py [hunk 4]
```diff
@@ -294,9 +295,47 @@ async def main(*, ruff1: Path, ruff2: Path, projects_jsonl: Optional[Path]) -> N
...
+                        current_changes = (current_changes[0] + 1, current_changes[1])
+                    elif line[0] == "-":
+                        current_changes = (current_changes[0], current_changes[1] + 1)
+
+                    rule_changes[rule_code] = current_changes
+
             else:
                 continue
 
+        print(f"Rules changed: {len(rule_changes.keys())}")
```
Comment (konstin) [inline scripts/check_ecosystem.py:326]: i'd only print this box if `rule_changes` is non-empty to shorten the comment a bit

File: scripts/check_ecosystem.py [hunk 5]
```diff
@@ -294,9 +295,47 @@ async def main(*, ruff1: Path, ruff2: Path, projects_jsonl: Optional[Path]) -> N
...
+        print()
+        print("| Rule | Changes | Additions | Removals |")
+        print("| ---- | ------- | --------- | -------- |")
+        for rule, (additions, removals) in dict(
+            sorted(
```
Comment (konstin) [inline scripts/check_ecosystem.py:331]: you can skip the `dict().items()` part and just iterated over the sorted tuples

**Review by konstin — COMMENTED** [review]
This is a good addition, i've left some code comments

File: scripts/check_ecosystem.py (hunk 4 above)
Comment (calumy) [reply scripts/check_ecosystem.py]: Addressed in 5400a18

File: scripts/check_ecosystem.py (hunk 2 above)
Comment (calumy) [reply scripts/check_ecosystem.py:303]: Addressed in 9a3c499fbbbcc3b422362dddcdfe40146e665146

File: scripts/check_ecosystem.py (hunk 3 above)
Comment (konstin) [reply scripts/check_ecosystem.py]: sorry, i wasn't precise enough: my idea was to put the rule into a group and then just use the group so you don't need to manually trim

Comment (calumy) [reply scripts/check_ecosystem.py]: Something more along the lines of 

```python
                for line in diff_str.splitlines():
                    # Find rule change for current line or construction
                    # + <rule>/<path>:<line>:<column>: <rule_code> <message>
                    matches = re.search(r": ([A-Z]{1,3}[0-9]{3,4})", line)

                    if matches is None:
                        # Handle case where there are no regex matches e.g.
                        # +                 "?application=AIRFLOW&authenticator=TEST_AUTH&role=TEST_ROLE&warehouse=TEST_WAREHOUSE" # noqa: E501, ERA001
                        # Which was found in local testing
                        continue

                    rule_code = matches.group(1)
```

I have pushed this in a commit, but due to the current [GitHub issues](https://www.githubstatus.com/), it does not appear to be showing up.

Comment (calumy) [reply scripts/check_ecosystem.py]: Updated in f9e09c5. I'm not sure why this is not showing up in this PR

Comment (konstin) [reply scripts/check_ecosystem.py]: yeah yesterday CI has also been flaky all day :/

thanks for fixing!

**Pull request merged**